File: cplusplus/Facade_Interface.cpp
/// @file
/// @brief
/// Implementation of the @ref DesignPatternExamples_cpp::IDeviceNetworkHighLevel "IDeviceNetworkHighLevel"
/// interface representing the high-level system used in the @ref facade_pattern "Facade pattern".

#include <memory>
#include "helpers/argumentnull_error.h"
#include "Facade_Interface.h"
#include "FacadeSubsystem_Interface.h"

namespace // Anonymous
{

    /// <summary>
    /// This class wraps the IDeviceNetworkLowLevel interface and implements
    /// the high level IDeviceNetworkHighLevel interface, which is a simpler
    /// interface.  All calls on the high level interface are forwarded to the
    /// appropriate low level interface.
    /// Part of the @ref facade_pattern "Facade pattern" example.
    /// </summary>
    class DeviceNetworkHighLevel : public DesignPatternExamples_cpp::IDeviceNetworkHighLevel
    {
    private:
        /// <summary>
        /// The low level system being wrapped by this class.
        /// </summary>
        DesignPatternExamples_cpp::IDeviceNetworkLowLevel* _lowlevelSystem;

    public:
        /// <summary>
        /// Constructor.
        /// </summary>
        /// <param name="system">The IDeviceNetworkLowLevel object to be
        /// wrapped by this class.</param>
        /// <exception cref="argumentnull_error">The system being wrapped
        /// cannot be null.</exception>
        DeviceNetworkHighLevel(DesignPatternExamples_cpp::IDeviceNetworkLowLevel* system)
            : _lowlevelSystem(system)
        {
            if (system == nullptr)
            {
                throw Helpers::argumentnull_error("system", "The system being wrapped cannot be null.");
            }
        }


        //====================================================================
        // IDeviceNetworkHighLevel methods
        //====================================================================

        /// @copydoc IDeviceNetworkHighLevel::NumChains()
        int NumChains() override
        {
            return _lowlevelSystem->GetNumChains();
        }

        /// @copydoc IDeviceNetworkHighLevel::GetIdcodes()
        std::vector<uint32_t> GetIdcodes(int chainIndex) override
        {
            std::vector<uint32_t> idcodes;

            if (_lowlevelSystem->LockDeviceChain(chainIndex))
            {
                idcodes = _lowlevelSystem->GetIdcodes(chainIndex);
                _lowlevelSystem->UnlockDeviceChain(chainIndex);
            }
            return idcodes;
        }

        /// @copydoc IDeviceNetworkHighLevel::EnableDevicesInDeviceChain()
        void EnableDevicesInDeviceChain(int chainIndex, uint32_t selectMask) override
        {
            if (_lowlevelSystem->LockDeviceChain(chainIndex))
            {
                _lowlevelSystem->EnableDevicesInDeviceChain(chainIndex, selectMask);
                _lowlevelSystem->UnlockDeviceChain(chainIndex);
            }
        }

        /// @copydoc IDeviceNetworkHighLevel::DisableDevicesInDeviceChain()
        void DisableDevicesInDeviceChain(int chainIndex) override
        {
            if (_lowlevelSystem->LockDeviceChain(chainIndex))
            {
                _lowlevelSystem->ResetDeviceChain(chainIndex);
                _lowlevelSystem->UnlockDeviceChain(chainIndex);
            }
        }
    };

    static std::unique_ptr<DeviceNetworkHighLevel> _instance;

} // end anonymous namespace


//########################################################################
//########################################################################


namespace DesignPatternExamples_cpp
{

    /////////////////////////////////////////////////////////////////////////
    // Function: CreateHighLevelInstance
    /////////////////////////////////////////////////////////////////////////
    IDeviceNetworkHighLevel* CreateHighLevelInstance()
    {
        if (!_instance)
        {
            IDeviceNetworkLowLevel* lowlevelSystem;
            lowlevelSystem = CreateLowLevelInstance();
            _instance = std::make_unique<DeviceNetworkHighLevel>(lowlevelSystem);
        }
        return _instance.get();
    }

} // end namespace

File: python/DesignPatternExamples_python/facade/facade_interface.py
## @file
#  @brief
#  Implementation of the
#  @ref DesignPatternExamples_python.facade.facade_interface.IDeviceNetworkHighLevel "IDeviceNetworkHighLevel"
#  interface representing the high-level system used in the @ref facade_pattern.

from abc import ABC, abstractmethod


## Represents a high level view of a complex network of device chains.
#  A device chain can be thought of as a list of devices that can be made
#  visible or hidden in the list but maintain the same relationship to
#  each other regardless of visibility.
#  
#  The methods on this high level interface may seem the same as on the
#  IDeviceNetworkLowLevel interface.  However, most of the methods on this
#  high level interface hide the need for taking a lock on a chain before
#  accessing the chain.
#  
#  This high level interface is a simplification and thus a facade for the
#  low level interface and the system underneath.
#  
#  Part of the @ref facade_pattern "Facade pattern" example.
class IDeviceNetworkHighLevel(ABC):
    ## The number of device chains available from the sub-system.
    @abstractmethod
    def NumChains(self) -> int:
        pass

    ## Returns a list of all idcodes from all selected devices in the
    #  given device chain.
    #
    #  @param chainIndex
    #         Index of the device chain to access (0..NumChains-1).
    #  @returns
    #     An array holding the idcodes of all visible devices in the given
    #  chain.
    @abstractmethod
    def GetIdcodes(self, chainIndex: int) -> list:
        pass

    ## Make visible certain devices in the given device chain.  The selectMask value
    #  has a bit set for each TAP device to make visible.
    #
    #  @param chainIndex
    #         Index of the device chain to access (0..NumChains-1).
    #  @param selectMask
    #         A bit mask where each bit corresponds to a TAP, up to the number
    #         of devices in the given device chain.  Bit 0 is ignored as the
    #         first TAP is always visible.
    @abstractmethod
    def EnableDevicesInDeviceChain(self, chainIndex : int, selectMask: int) -> None:
        pass

    ## Resets the given device chain so that all devices except the TAP
    #  controller is no longer visible.
    #
    #  @param chainIndex
    #         Index of the device chain to access (0..NumChains-1).
    @abstractmethod
    def DisableDevicesInDeviceChain(self, chainIndex : int) -> None:
        pass

